File: in-js/vm.js
const CODE = [
  'PUSH', 5,
  'PUSH', 4,
  'ADD',
  'PUSH', 3,
  'MINUS'
]

function VM(codes) {
  const stack = []
  let cI = 0
  let sI = 0
  const len = codes.length
  while(cI < len) {
    const curCode = codes[cI]
    console.log('curCode==', curCode)
    switch(curCode) {
      case 'PUSH':
        // 入栈
        stack[sI] = codes[cI + 1]
        cI++;
        cI++;
        sI++;
        break
      case 'ADD':
        // 出栈
        stack[sI] = stack[sI - 1] + stack[sI - 2]
        cI++;
        sI++;
        break
      case 'MINUS':
        // 出栈
        stack[sI] = stack[sI - 2] - stack[sI - 1]
        cI++;
        sI++;
        break
    }
  }
  console.log(stack)
}

VM(CODE)

File: in-go/vm.go
package main

import (
	"fmt"
	"strconv"
)

func VM(codes []string) {
	cLen := len(codes)
	stack := make([]string, cLen)
	cI := 0
	sI := 0
	for cI < cLen {
		curCode := codes[cI]
		switch curCode {
		case "PUSH":
			stack[sI] = codes[cI+1]
			cI++
			sI++

		case "ADD":
			right, _ := strconv.Atoi(stack[sI-1])
			sI--
			left, _ := strconv.Atoi(stack[sI-1])
			sI--
			stack[sI] = strconv.Itoa(right + left)
			sI++

		case "MINUS":
			right, _ := strconv.Atoi(stack[sI-1])
			sI--
			left, _ := strconv.Atoi(stack[sI-1])
			sI--
			stack[sI] = strconv.Itoa(left - right)
			sI++

		}
		cI++
	}
	fmt.Println(stack[sI-1])
}
func main() {
	CODES := []string{
		"PUSH", "4",
		"PUSH", "5",
		"ADD",
		"PUSH", "3",
		"MINUS",
	}
	fmt.Println("测试")
	VM(CODES)
}
